### DISCUSSION TITLE
Having a tough time with `ebitenmobile`

### DISCUSSION THREAD

**Discussion by MilanKladivko** [hajimehoshi/ebiten#1856] [Q&A] [answered]
**I must be doing something wrong, spent a few hours with no outcome...**

I tried to follow the *very basic* steps in [https://ebiten.org/documents/mobile.html](). Installed `ebitenmobile`, tried to run it in a non-Go-project folder, got the following error back... 

```
~/dekstop >  ebitenmobile bind -target android -javapkg test.mobile -o ~/desktop/out.aar .
2021/10/27 21:56:41 go [build -o bin/gomobile golang.org/x/mobile/cmd/gomobile] failed: /home/mikl/.go/pkg/mod/golang.org/x/mobile@v0.0.0-20210902104108-5d9a33257ab5/cmd/gomobile/bind.go:19:2: missing go.sum entry for module providing package golang.org/x/mod/modfile (imported by golang.org/x/mobile/cmd/gomobile); to add:
	go get golang.org/x/mobile/cmd/gomobile@v0.0.0-20210902104108-5d9a33257ab5
/home/mikl/.go/pkg/mod/golang.org/x/tools@v0.1.2/internal/gocommand/vendor.go:17:2: missing go.sum entry for module providing package golang.org/x/mod/semver (imported by golang.org/x/tools/internal/gocommand); to add:
	go get golang.org/x/tools/internal/gocommand@v0.1.2
/home/mikl/.go/pkg/mod/golang.org/x/tools@v0.1.2/go/packages/golist.go:29:2: missing go.sum entry for module providing package golang.org/x/xerrors (imported by golang.org/x/tools/go/packages); to add:
	go get golang.org/x/tools/go/packages@v0.1.2

exit status 1
```

Then I tried probably all combinations of installing these inside/outside of various projects, `go get` vs `go install` vs `go get -d` ... 
Nothing changed throughout all off these, so I'm thinking it's not about what is / isn't installed or where. ...?

Any ideas? What am I not seeing? Anything besides trying other versions of Go that I should try?

*( Running Go1.16 on Linux Ubuntu 20.04 if that matters )*

**Comment by hajimehoshi** [comment, marked as answer]
Which version of Ebiten are you using? Thanks,

**Reply by hajimehoshi** [reply]
I'm not sure but my guess is that Ebiten's version and Ebitenmobile's version don't match

**Reply by hajimehoshi** [reply]
To force to match them, you can do:

```
go run github.com/hajimehoshi/ebiten/v2/cmd/ebitenmobile bind -target android -javapkg test.mobile -o ~/desktop/out.aar .
```

assuming that the current directory has `go.mod`.

**Reply by MilanKladivko** [reply]
That does seem to make sense -- that would use the code presumably already living in the ebiten package used specifically for the module to begin with. 

But what I then get seems like unresolved dependencies for it to me? It complained more about go.sum entries, starting with Go saying "please install that `ebitenmobile` thing correctly". So I did, but then it gets stuck again -- but now on a different set of packages. It's baffling to me... 

Here's all the steps I did after deleting my `go.sum` and `go.mod` followed by `go mod tidy` and trying again. 

```
mikl ~/work/test-ebiten-mobile 
 > go run github.com/hajimehoshi/ebiten/v2/cmd/ebitenmobile bind -target android -javapkg test.mobile -o ~/desktop/out.aar ./mobile
../../.go/pkg/mod/github.com/hajimehoshi/ebiten/v2@v2.2.1/cmd/ebitenmobile/main.go:34:2: missing go.sum entry for module providing package golang.org/x/tools/go/packages (imported by github.com/hajimehoshi/ebiten/v2/cmd/ebitenmobile); to add:
	go get github.com/hajimehoshi/ebiten/v2/cmd/ebitenmobile@v2.2.1

mikl ~/work/test-ebiten-mobile 
 > go get -d github.com/hajimehoshi/ebiten/v2/cmd/ebitenmobile@v2.2.1

mikl ~/work/test-ebiten-mobile 
 > cat go.mod
module testmobile

go 1.16

require github.com/hajimehoshi/ebiten/v2 v2.2.1

mikl ~/work/test-ebiten-mobile 
 > go run github.com/hajimehoshi/ebiten/v2/cmd/ebitenmobile bind -target android -javapkg test.mobile -o ~/desktop/out.aar ./mobile
../../.go/pkg/mod/golang.org/x/tools@v0.1.6/internal/gocommand/vendor.go:17:2: missing go.sum entry for module providing package golang.org/x/mod/semver (imported by golang.org/x/tools/internal/gocommand); to add:
	go get golang.org/x/tools/internal/gocommand@v0.1.6
../../.go/pkg/mod/golang.org/x/tools@v0.1.6/go/packages/golist.go:29:2: missing go.sum entry for module providing package golang.org/x/xerrors (imported by golang.org/x/tools/go/packages); to add:
	go get golang.org/x/tools/go/packages@v0.1.6

mikl ~/work/test-ebiten-mobile 
 > go get -d golang.org/x/tools/internal/gocommand@v0.1.6

mikl ~/work/test-ebiten-mobile 
 > go get -d golang.org/x/tools/go/packages@v0.1.6

mikl ~/work/test-ebiten-mobile 
 > cat go.mod
module testmobile

go 1.16

require (
	github.com/hajimehoshi/ebiten/v2 v2.2.1
	golang.org/x/tools v0.1.6 // indirect
)

mikl ~/work/test-ebiten-mobile 
 > go run github.com/hajimehoshi/ebiten/v2/cmd/ebitenmobile bind -target android -javapkg test.mobile -o ~/desktop/out.aar ./mobile
../../.go/pkg/mod/golang.org/x/tools@v0.1.6/internal/gocommand/vendor.go:17:2: missing go.sum entry for module providing package golang.org/x/mod/semver (imported by golang.org/x/tools/internal/gocommand); to add:
	go get golang.org/x/tools/internal/gocommand@v0.1.6
../../.go/pkg/mod/golang.org/x/tools@v0.1.6/go/packages/golist.go:29:2: missing go.sum entry for module providing package golang.org/x/xerrors (imported by golang.org/x/tools/go/packages); to add:
	go get golang.org/x/tools/go/packages@v0.1.6

```

**Reply by hajimehoshi** [reply]
> ../../.go/pkg/mod/github.com/hajimehoshi/ebiten/v2@v2.2.1/cmd/ebitenmobile/main.go:34:2: missing go.sum entry for module providing package golang.org/x/tools/go/packages (imported by github.com/hajimehoshi/ebiten/v2/cmd/ebitenmobile); to add:

This complains that `go.sum` doesn't have enough information when building `gomobile` at a temporary directory, so it was looking at `go.sum` in the temporary directory. That's pretty odd since `go mod tidy` is executed before building. I'll take a look later. Thanks,

**Reply by hajimehoshi** [reply]
What will be output with `-x`?

```
go run github.com/hajimehoshi/ebiten/v2/cmd/ebitenmobile bind -target android -javapkg test.mobile -o ~/desktop/out.aar -x .
```

**Reply by MilanKladivko** [reply]
Same thing as without the `-x`.

**Reply by hajimehoshi** [reply]
`-x` should show the log. Can I take a look?

**Reply by hajimehoshi** [reply]
Ah, I realized that the error is in building ebitenmobile itself...: https://github.com/hajimehoshi/ebiten/blob/main/cmd/ebitenmobile/main.go#L34

So my assumption was wrong. This complains go.sum in Ebiten lacks some information. However, I'm still not sure why...

**Reply by hajimehoshi** [reply]
What about Go 1.17?

**Reply by hajimehoshi** [reply]
Also, I'd like to see your go.mod in your local directory.

**Reply by MilanKladivko** [reply]
```
 > cat go.mod
module testmobile

go 1.16

require (
	github.com/hajimehoshi/ebiten/v2 v2.2.1
	golang.org/x/tools v0.1.6 // indirect
)
```

I've tried to copy the project to another computer with Go 1.17.1 installed. After modifying the Go version in the `go.mod` file to 1.17 and running `go get -d [...]/ebitenmobile`, the long-path command ran for a while and got stuck on an Android NDK missing error -- I don't have Android Studio installed on this computer. 
Then my internet went down, so downloading AndroidSDK+other deps and testing the rest of the process is not an option, *but in the meantime...*

Is the original issue, *"ebitenmobile doesn't even build on 1.16 for go.sum reasons"*, resolved to *"use 1.17 instead"*? 

Running with "-x" now prints PATH and so on, so it must have built `ebitenmobile` correctly this time and I'm getting runtime errors, which are valid (no internet, cannot go get). 
As far as I'm concerned, my issue is solved but I'll report back when I produce a full build to make sure.

**Reply by hajimehoshi** [reply]
OK so the Go version was the problem. The error message was pretty unfriendly. It's great to report this issue to the Go team.

On the other hand, Go 1.16 should succeed to build ebitenmobile: https://github.com/hajimehoshi/ebiten/runs/4013690738?check_suite_focus=true so this seems pretty odd...

**Reply by MilanKladivko** [reply]
Alright! Got the `.aar` file on Go1.17.2, after painful PATH setting, installing the NDK, more PATH setting and installing the 3GB+ of Android stuff. 

Gotta say, making this work is not pleasant... But I can't imagine how long it took to figure out how to make this work to begin with. I'd like to see a few words on the web page about all this setup (including setting PATHs and what one needs to install or how big it's gonna be) -- but in the end I figured it out without that, so no hard feelings. I'm just glad it actually works!!

Thank you very much for your quick responses!
